### ISSUE TITLE
Flavor Naming Scheme: Use of colon character as delimiter

### ISSUE THREAD

**Issue by wwentland** [SovereignCloudStack/standards#190]
# Summary

The current [flavor naming standard][scs-flavor-naming] uses colon characters (i.e. `:`) as delimiters. This is an unfortunate choice, as it is incompatible with requirements stipulated by downstream applications such as Kubernetes.

# Details

Kubernetes uses [labels][k8s-labels] and [annotations][k8s-annotations] for a variety of use cases. Most importantly in the context of this discussion will be their use for [taints and tolerations][k8s-taints-tolerations] and the standard [well-known labels, annotations and taints][k8s-well-known-labels-annotations-taints].

The requirements for valid [label][k8s-label-charset] values are:

- must be 63 characters or less (can be empty),
- unless empty, must begin and end with an alphanumeric character ([a-z0-9A-Z]),
- could contain dashes (-), underscores (_), dots (.), and alphanumerics between.

That is, it must adhere to the regular expression `(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')`.

The usage of colon characters is not allowed and their usage in flavor names causes issues when, for example, the kubelet wants to set the well-known `node.kubernetes.io/instance-type` label:

```
Jan 01 08:09:10 foo-bar-box hyperkube[12345]: I0103 08:09:10.23456 14821 kubelet_node_status.go:426] "Adding node label from cloud provider" labelKey="node.kubernetes.io/instance-type" labelValue="SCS-4V:16:50"
[...]
Jan 01 08:09:10 foo-bar-box hyperkube[12345]: E0103 08:09:10.23456 14821 kubelet_node_status.go:94] "Unable to register node with API server" err="Node \"foo-bar-node\" is invalid: metadata.labels: Invalid value: \"SCS-4V:16:50\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue', or 'my_value', or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')" node="foo-bar-node"
```

Given the significance of Kubernetes in the SCS project, it would be advisable to reconsider the naming scheme and adopt one that is compatible with the data model defined there.

Thank you!

[scs-flavor-naming]: https://github.com/SovereignCloudStack/standards/blob/main/Design-Docs/flavor-naming.md
[k8s-labels]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
[k8s-annotations]: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
[k8s-taints-tolerations]: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
[k8s-well-known-labels-annotations-taints]: https://kubernetes.io/docs/reference/labels-annotations-taints/
[k8s-label-charset]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

**Comment by garloff** [comment]
Options (theoretical, without ordering):
0) Ignore OpenShift
1) See whether OpenShift can handle this differently (e.g. translate flavor names)
2) See whether k8s can be changed to be more tolerant
3) See whether flavor spec needs adjustment
4) Reevaluate whether we need flavor spec

**Comment by garloff** [comment]
Are any other IaaS resource types from IaaS potentially used 1:1 in k8s and need thus to conform to k8s annotation rules?
- Storage classes?

**Comment by garloff** [comment]
@wwentland - I have not seen OpenStack flavor names ending up in k8s annotations.
@janiskemper, @batistein - have you?
It seems that this is not all that common.

However, OpenShift may have good reasons ...
- Do we understand why it's doing this? Tolerations? Taints?
- Can we tweak the way OpenShift is used to not push flavor names into k8s annotations? Or quote/translate them?
- How likely are other tools doing something similar?

**Comment by batistein** [comment]
This is definitely a problem when the ccm tries to use the flavor name for the label. In pluscloud-based clusters, the ccm use: `node.kubernetes.io/instance-type: 06894112-a4b4-466b-a2ba-38faf311ae5c`. 
So the problem does not occur there. But of course, the type of the label is not helpful when you want to do scheduling.
In all other Kubernetes clusters I know, the type of the node is mapped 1:1!
So I would agree and state that the current flavor scheme is not RFC 1123 compliant, which is a requirement for use in Kubernetes.

**Comment by janiskemper** [comment]
In the English documentation they say that the node controller of the CCM annotates and labels the node object with cloud-specific information (https://kubernetes.io/docs/concepts/architecture/cloud-controller/#node-controller). In the German version, they even explicitly mention the type of the cloud instance (https://kubernetes.io/de/docs/concepts/architecture/cloud-controller/#node-controller).

**Comment by fkr** [comment]
> So I would agree and state that the current flavor scheme is not RFC 1123 compliant, which is a requirement for use in Kubernetes.

Thanks @batistein - that is a very helpful bit of info!

**Comment by garloff** [comment]
Hmm, it thus seems we should seriously consider changing our spec, accepting (and mitigating as far as possible) the pain this causes to our already existing flavor spec adopters. Input on #192 is very welcome!

**Comment by garloff** [comment]
Did some more reading.
k8s is actually with annotations containing colons. But not labels. Labels follow stricter rules as written in the report.
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
So adjusting would seem unavoidable.

**Comment by garloff** [comment]
This has been addressed in the flavor-naming-scheme v2 (scs-0110-v2), merged with PR #226.
